### PR TITLE
Implementation of restart for a GenericContainer

### PR DESCRIPTION
**Pull request by corbs9**
- Refactored `start()` to call `tryRun()` which calls `tryCreate()` followed by `tryStart()` to reuse the start functionality for the restart method.
- Implemented functionality to allow a running container to be restarted, via docker stop followed by a docker start.

### CODE REVIEW

**Comment by bsideup** [comment]
Fixes #616

**Comment by corbs9** [comment]
Hi @bsideup, 

I think there is a problem with some of the containers which are being used in some of the other tests. 

```
$ docker pull httpd:2.4-alpine
2.4-alpine: Pulling from library/httpd
no matching manifest for linux/amd64 in the manifest list entries
```

These are causing a few tests to fail in `org.testcontainers.spock.TestcontainersSharedContainerIT`

Would you like me to take a look at this as part of this PR?

**Comment by rnorth** [comment]
@corbs9 I encountered exactly the same image problem at probably the same time as you did. It seems to have been fixed by the image maintainers, so no need for you to look into that.

We will soon be able to pin our image versions to specific sha256 hashes, which might help avoid such problems in the future.

**Comment by kiview** [comment]
This can also occur if you are loosing network connectivity during some stage of pulling the image, or sometimes just as a Docker Hub hickup, happens sometimes in our CI as well, completly unrelated to testcontainers (i.e. Java image).

**Comment by rnorth** [comment]
I think we're good to go with this, are we not, @bsideup ? I'll just fix the changelog merge conflict...

**Comment by bsideup** [comment]
@rnorth let me review it, missed this one, but there are some areas to improve (on a first look)

File: core/src/test/java/org/testcontainers/junit/GenericContainerRuleTest.java [hunk 1]
```diff
@@ -400,4 +401,27 @@ public void addExposedPortAfterWithExposedPortsTest() {
         assertTrue("withExposedPort should be exposed", redis.getExposedPorts().contains(REDIS_PORT));
         assertTrue("addExposedPort should be exposed", redis.getExposedPorts().contains(8987));
     }
+
+    @Test
+    public void restartShouldKeepTheSameContainerButUpdateStartedTime() {
```
Comment (bsideup) [inline core/src/test/java/org/testcontainers/junit/GenericContainerRuleTest.java]: please use `try {} finally {}` and stop the containers after the test

File: core/src/main/java/org/testcontainers/containers/GenericContainer.java [hunk 2]
```diff
@@ -206,20 +206,75 @@ public void start() {
...
+    /**
+     * Restarts a running container.
+     *
+     * @throws ContainerNotStartedException if {@link #start()} hasn't been called first
+     */
+    public void restart() {
+        Profiler profiler = new Profiler("Container restart");
+        profiler.setLogger(logger());
+
+        if (null == containerId) {
```
Comment (bsideup) [inline core/src/main/java/org/testcontainers/containers/GenericContainer.java]: I would remove this check, IMO It's better for UX to not fail here and just start a new container

@rnorth WDYT?

File: core/src/main/java/org/testcontainers/containers/GenericContainer.java [hunk 3]
```diff
@@ -206,20 +206,75 @@ public void start() {
...
+                    attempt.incrementAndGet(), startupAttempts);
+                tryRestart(profiler.startNested("Container restart attempt"));
+                return true;
+            });
+        } finally {
+            profiler.stop().log();
+        }
+    }
+
+    private void tryRestart(Profiler profiler) {
```
Comment (bsideup) [inline core/src/main/java/org/testcontainers/containers/GenericContainer.java]: I don't think we need these private methods (`tryRestart`, `tryRun`), they make code looks like there is some logic behind it

File: core/src/main/java/org/testcontainers/containers/GenericContainer.java [hunk 4]
```diff
@@ -228,7 +283,12 @@ private void tryStart(Profiler profiler) {
             applyConfiguration(createCommand);
 
             containerId = createCommand.exec().getId();
```
[the remark this answers is not in the document]
Comment (corbs9) [reply core/src/main/java/org/testcontainers/containers/GenericContainer.java]: I believe this was a pre-existing side effect and not something that was brought in by this change. Would you like me to change it as part of this PR?

File: core/src/main/java/org/testcontainers/containers/GenericContainer.java (hunk 2 above)
Comment (corbs9) [reply core/src/main/java/org/testcontainers/containers/GenericContainer.java]: I can do that, it's not a problem. @rnorth please let me know if you'd prefer this functionality.

File: core/src/main/java/org/testcontainers/containers/GenericContainer.java [hunk 5]
```diff
@@ -206,20 +206,75 @@ public void start() {
...
         try {
-            String dockerImageName = image.get();
+            profiler.start("Restarting running container");
+            logger().debug("Restarting container: {}", getDockerImageName());
+
+            AtomicInteger attempt = new AtomicInteger(0);
+            Unreliables.retryUntilSuccess(startupAttempts, () -> {
+                logger().debug("Trying to restart container: {} (attempt {}/{})", image.get(),
+                    attempt.incrementAndGet(), startupAttempts);
+                tryRestart(profiler.startNested("Container restart attempt"));
```
[the remark this answers is not in the document]
Comment (corbs9) [reply core/src/main/java/org/testcontainers/containers/GenericContainer.java]: I'm not too sure what you mean by this. Most of the code is the same as it was before and I have split out methods into smaller, focussed methods and reused functionality.

**Comment by bsideup** [comment]
Hi @corbs9,
Sorry for taking so long to review it, I'll go thru the changes you've made after the last review soon 👍

**Comment by bsideup** [comment]
@corbs9 FYI I've fixed the merge conflict

File: core/src/main/java/org/testcontainers/containers/GenericContainer.java (hunk 2 above)
Comment (bsideup) [reply core/src/main/java/org/testcontainers/containers/GenericContainer.java]: @rnorth ping

**Comment by ayedo** [comment]
Hi all, first of all, thanks for your great work!

I would really like to have this functionality, and might be able to help out?

Is there anything left to do for this pull-request to be merged?

**Comment by chungngoops** [comment]
@ayedo : I am working on supporting `restart` container as the discussion in the #606

**Pull request closed without merging**